02_bubble/willshen8/bubble.go

```diff
@@ -0,0 +1,68 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func bubble(s []int) []int {
+	sliceCopy := append(s[:0:0], s...)
+	sliceLength := len(s)
+	sorted := false
+
+	for !sorted {
+		swapOccurred := false
+		for i := 0; i < sliceLength-1; i++ {
+			if sliceCopy[i+1] < sliceCopy[i] {
+				swap(sliceCopy, i, i+1)
+				swapOccurred = true
+			}
+		}
+		if !swapOccurred {
+			sorted = true
+		}
+	}
+	return sliceCopy
+}
+
+func insertionSort(s []int) []int {
+	sliceCopy := append(s[:0:0], s...)
+	for i := 1; i < len(s); i++ {
+		currValue := sliceCopy[i]
+		emptySlot := i
+
+		for emptySlot > 0 && sliceCopy[emptySlot-1] > currValue {
+			sliceCopy[emptySlot] = sliceCopy[emptySlot-1]
+			emptySlot--
+		}
+		sliceCopy[emptySlot] = currValue
+
+	}
+	return sliceCopy
+}
+
+func swap(slice []int, pos1 int, pos2 int) []int {
+	temp := slice[pos2]
+	slice[pos2] = slice[pos1]
+	slice[pos1] = temp
+	return slice
+}
+
+func equal(a, b []int) bool {
```

Comment:
Have a look at testify/assert, it was recommended to me by @juliaogris and spares you from needing to implement your own slice equality (and diff printing) code.

Reply:
Great point, that's for pointing that out for me. Will use testify next time.